src/Cardano/CoinSelection/LargestFirst.hs

```diff
@@ -20,77 +21,140 @@ import Cardano.Types
 import Control.Arrow
     ( left )
 import Control.Monad
-    ( foldM, when )
+    ( foldM )
 import Control.Monad.Trans.Except
     ( ExceptT (..), except, throwE )
 import Data.Functor
     ( ($>) )
 import Data.List.NonEmpty
     ( NonEmpty (..) )
 import Data.Ord
-    ( Down (..), comparing )
+    ( Down (..) )
 
 import qualified Data.List as L
 import qualified Data.List.NonEmpty as NE
 import qualified Data.Map.Strict as Map
 
--- | Implements the __Largest-First__ coin selection algorithm.
+-- | Generate a coin selection according to the __largest first__ algorithm.
+--
+-- For the given /output list/ and /starting UTxO set/, this algorithm
+-- generates a /coin selection/ that is capable of paying for all of the
+-- outputs, and a /remaining UTxO set/ from which spent values have been
+-- removed.
+--
+-- === Cardinality Rules
+--
+-- The algorithm requires that:
+--
+--  1.  Each output is paid for by /one or more/ entries from the UTxO set.
+--
+--  2.  Each entry from the UTxO set is used to pay for /at most one/ output.
+--
+--      (A given entry from the UTxO set /cannot/ be used to pay for multiple
+--      outputs.)
+--
+-- === Order of Processing
+--
+-- The algorithm processes /both/ the given output list /and/ the supplied UTxO
+-- set in __descending order of coin value__, from largest to smallest.
+--
+-- At all stages of processing, the algorithm maintains a /remaining UTxO set/
+-- that is steadily depleted as outputs are paid for.
+--
+-- === Processing an Output
+--
+-- For /each output/ in the (descending) output list, the algorithm repeatedly
+-- selects unspent values from the /remaining UTxO set/ (in descending order)
+-- until the /total selected value/ is greater than (or equal to) the output
+-- value, at which point the algorithm move ons to processing the /next/
+-- output.
+--
+-- If the /total selected value/ is greater than required for a particular
+-- output, the algorithm generates a /change output/ with the exact difference
+-- in value.
+--
+-- === Termination Conditions
+--
+-- The algorithm terminates with an __error__ if:
+--
+--  1.  The /total value/ of the starting UTxO set (the amount of money
+--      /available/) is /less than/ the total value of the output list (the
+--      amount of money /required/).
+--
+--      See: __'ErrUtxoBalanceInsufficient'__.
+--
+--  2.  The /number/ of entries in the starting UTxO set is /smaller than/ the
+--      number of requested outputs.
+--
+--      Due to the nature of the algorithm, /at least one/ UTxO entry is
+--      required /for each/ output.
+--
+--      See: __'ErrUtxoNotFragmentedEnough'__.
+--
+--  3.  Due to the particular /distribution/ of values within the starting UTxO
+--      set, the algorithm depletes all entries from the set /before/ it is able
```

Comment:
maybe you meant : `and it is unable to cover all requested outputs` ?

Reply:
In this case, I think the wording makes sense.

I tend to view this algorithm as a **race** between:

- Depleting the remaining UTxO
- Depleting the remaining list of unpaid outputs

We **ideally** want to deplete the list of unpaid outputs **before** we deplete the remaining UTxO.

The failure case occurs when we deplete the remaining UTxO before we are able to deplete the list of unpaid outputs, i.e., _before we are able to pay for all the outputs_.

I do think the current phrasing is correct, but I'll have another think about whether there's a better way to phrase this.

Reply:
@paweljakubas I've updated the phrasing with this change: 

https://github.com/input-output-hk/cardano-coin-selection/pull/7/commits/973126e81272aaef88f4875f7d6a75c5956ea95e